### PR TITLE
fix: excessive cloning in `runSubscribersWithMutation`

### LINKED ISSUE #883: Performance Issue: excessive `structuredClone` in `runSubscribersWithMutation`
The `runSubscribersWithMutation` function has a critical performance bottleneck causing long execution time with the potential to hang the browser completely.

## Impact

- The example dojo app has 26 active subscribers during streaming
- Message arrays grow from 0 to 42 messages during a 742KB test run with ~200 events
- Current execution time: 810ms

| Traffic                                                                                                                              | Events                                                                                                                               | Trace                                                                                                                                |
| ------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------ |
| Image | Image | Image |

## Root Cause

This is in the hot path, called 2× per streaming event (once for `onEvent`, once for specific event handler), but it:

- Clones per subscriber even if they do not mutate
- Does expensive change detection via `JSON.stringify()`

## Proposed Solution

I have a working fix on my fork with two commits:

- Move cloning from inputs to outputs https://github.com/scoutqa-dot-ai/ag-ui/commit/27b7ab28fe7da6a78f3009f227cb50129f0f1fc5
- Use reference equality to identify changes https://github.com/scoutqa-dot-ai/ag-ui/commit/bdb380e29480d2a873610796932f1fa3dc8f3ecc

## Expected Results

Image

- New execution time: ~53ms (15x faster)
- Eliminate serialization bottleneck visible in flame graphs

I'm ready to submit a PR once this approach is validated.

### PR DESCRIPTION
**Pull request by daohoangson**
This PR fixes #883

The `runSubscribersWithMutation` function has a critical performance bottleneck causing long execution time with the potential to hang the browser completely.

## Impact

- The example dojo app has 26 active subscribers during streaming
- Message arrays grow from 0 to 42 messages during a 742KB test run with ~200 events
- Current execution time: 810ms

| Traffic                                                                                                                              | Events                                                                                                                               | Trace                                                                                                                                |
| ------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------ |
| Image | Image | Image |

## Root Cause

This is in the hot path, called 2× per streaming event (once for `onEvent`, once for specific event handler), but it:

- Clones per subscriber even if they do not mutate
- Does expensive change detection via `JSON.stringify()`

## The Fix

- Move cloning from inputs to outputs
- Use reference equality to identify changes

## Expected Results

Image

- New execution time: ~53ms (15x faster)
- Eliminate serialization bottleneck visible in flame graphs

### CODE REVIEW

File: sdks/typescript/packages/client/src/agent/subscriber.ts [hunk 1]
```diff
@@ -178,18 +178,16 @@ export async function runSubscribersWithMutation(
...
   let stopPropagation: boolean | undefined = undefined;
 
   for (const subscriber of subscribers) {
     try {
-      const mutation = await executor(
-        subscriber,
-        structuredClone_(messages),
-        structuredClone_(state),
-      );
+      const mutation = await executor(subscriber, messages, state);
```
[the remark this answers is not in the document]
Comment (daohoangson) [reply sdks/typescript/packages/client/src/agent/subscriber.ts:250]: This is intentional.

**Comment by daohoangson** [comment]
The failing job is unrelated to the change in this PR, reported separately here https://github.com/ag-ui-protocol/ag-ui/issues/897

**Comment by mattbasta** [comment]
Thanks for submitting this fix!

**Review by jpr5 — CHANGES_REQUESTED** [review]
## PR #899 Review — ag-ui-protocol/ag-ui (2026-03-15)

### PR #899: fix: excessive cloning in `runSubscribersWithMutation`

**Author:** daohoangson | **Status:** Open | **Base:** main ← perf/structured-clone **File changed:** `sdks/typescript/packages/client/src/agent/subscriber.ts`

---

### Summary

This PR optimizes `runSubscribersWithMutation` — a hot-path function called 2x per streaming event with \~26 active subscribers — by:

1. Moving `structuredClone_` from inputs (per-subscriber) to outputs (only on mutation)
2. Replacing `JSON.stringify` equality checks with reference equality

Claimed improvement: \~810ms → \~53ms (15x faster). The optimization is sound in principle but introduces a **semantic contract change** that removes per-subscriber isolation.

---

### Critical Issues

#### 1. In-place mutations by subscribers silently corrupt shared state and evade change detection

**File:** `subscriber.ts:190, 224-226`

Previously, each subscriber received a `structuredClone_` copy, making in-place mutations harmless. Now subscribers share the same mutable reference. If a subscriber mutates `messages` or `state` in-place (e.g., `messages.push(...)`) and returns `undefined`, the mutation:

- **Corrupts** `messages0`**/**`state0` (the baseline clone), because `messages` still points to the same object
- **Leaks to all subsequent subscribers** without going through the mutation return path
- **Evades change detection** at line 225: `messages !== messages0` is `false` (same reference), so the return value reports no change even though the data was modified

This is a **public API contract change** in a library. Downstream consumers who relied on subscriber isolation will silently break.

**Concrete scenario:**

```typescript
const subscriberA = {
  onEvent({ messages }) {
    messages.push(newMessage); // in-place, returns void
  }
};
// subscriberB now sees the pushed message (new behavior)
// But the return value has no `messages` field — mutation is silently lost
```

**Recommendation:** Either (a) add `Object.freeze()` in dev mode to catch violations, (b) use `ReadonlyArray<Message>` in the executor signature for compile-time friction, or (c) restore per-subscriber cloning while keeping the reference-equality optimization on the return path (which is the bigger win anyway).

#### 2. Broad catch block swallows all errors, completely silent in test environments

**File:** `subscriber.ts:211-221` (pre-existing, worsened by this PR)

The catch block catches all errors from subscriber execution, including `structuredClone_` failures on mutation output (lines 199/203). In test environments, errors are completely suppressed. With the new code, a `DataCloneError` from non-cloneable mutation values will be silently swallowed, the mutation lost, and subsequent subscribers will see stale data.

This is pre-existing but now worse because `structuredClone_` calls moved inside the try block.

---

### Important Issues

#### 3. `structuredClone_` fallback silently produces shallow copies

**File:** `utils.ts:8-12` (pre-existing)

The `structuredClone_` utility falls back to `{ ...obj }` (shallow spread) when `JSON.stringify` throws. This is not a deep clone — nested objects still share references. With the PR's clone-on-output strategy, a shallow copy means nested message properties (like `toolCalls` arrays) would still be shared between the cloned value and the original, defeating the purpose. The fallback has **no logging**.

#### 4. No tests for the changed isolation semantics

**File:** `subscriber.test.ts`

The PR adds no new tests despite changing a fundamental behavioral contract. There are no tests for:

- A subscriber that mutates inputs in-place without returning a mutation
- A subscriber that returns the same reference it received
- `structuredClone_` failure on mutation output
- The reference-equality vs. `JSON.stringify` semantic difference

The existing 31 tests all pass, but they only exercise the happy path where subscribers return new objects.

#### 5. Reference equality reports false positives compared to old behavior

**File:** `subscriber.ts:225-226`

If a subscriber returns `{ messages: structuredClone_(messages) }` (a copy with identical content), the new code reports a change (different reference) while the old `JSON.stringify` code would not. This could cause unnecessary `onMessagesChanged` / `onStateChanged` callbacks and UI re-renders.

This is a reasonable trade-off for 15x performance, but should be documented.

---

### Suggestions

#### 6. Rename `messages0`/`state0` for clarity

**File:** `subscriber.ts:181-182`

The `0` suffix is uncommon in TypeScript and ambiguous (index? version? original?). Consider `baselineMessages`/`baselineState` or `originalMessages`/`originalState`.

#### 7. Add a comment documenting the no-in-place-mutation contract

**File:** `subscriber.ts:190`

Above the executor call, add a comment explaining that subscribers receive shared references and must not mutate them in-place. The old code's defensive cloning made this implicit contract invisible.

#### 8. Fix misleading "Merge" comment

**File:** `subscriber.ts:197`

The comment says "Merge messages/state" but the code performs a full replacement with a defensive clone. Consider: `// Replace with a defensive copy of the subscriber's mutation`.

#### 9. Fix misleading catch block comment

**File:** `subscriber.ts:219`

The comment `// Continue to next subscriber unless we want to stop propagation` implies conditional behavior, but the `continue` always executes. Rewrite to `// Skip this subscriber's mutation and continue`.

---

### Strengths

- **Significant, measurable performance win** — 15x improvement (810ms → 53ms) on a real-world workload with 26 subscribers and 200+ events. Well-profiled with flame graphs.
- **Correct optimization strategy** — Cloning on output rather than input is sound when most subscribers don't mutate. Clone count drops from O(N) per event to O(mutations) per event.
- **Reference equality is a genuine improvement** over `JSON.stringify` for change detection — cheaper, no serialization cost, avoids edge cases with non-JSON-safe values.
- **Small, focused diff** — Single file, \~20 lines changed. Easy to review and reason about.
- **All 31 existing tests pass** with the new semantics, providing confidence for the well-behaved subscriber path.

---

### Requested Changes

The performance optimization is valuable and the core approach is sound, but the PR should address:

1. **Add at least one test** documenting the new contract (subscribers must not mutate inputs in-place), or testing the behavior when they do
2. **Add compile-time protection** via `ReadonlyArray<Readonly<Message>>` and `Readonly<State>` on the executor parameter types
3. **Add a brief comment** above line 190 documenting that subscribers receive shared references
4. **Consider** `Object.freeze` **in dev mode** as a runtime guard against the most dangerous failure mode

**Comment by daohoangson** [comment]
Thanks for the review, I'll update the PR as suggested.

**Comment by daohoangson** [comment]
Thanks for your comprehensive review, I have pushed a few commits:

- Added a few tests to verify the isolation contract
- Added compile time protection as suggested
- Fixed outdated comments
- Added deep freeze for dev/test environments

**Comment by magalhas** [comment]
How long until this gets reviewed and merged? I'd like very much to get rid of the patch I'm using on the npm package and update the library. The client is unusable without this fix.

**Comment by daohoangson** [comment]
Yeah, would love to have this merged.

**Comment by jpr5** [comment]
Opened #1395 which includes all of @daohoangson's original commits plus review fixes (2 additional commits addressing the feedback from the code review). The fork's `maintainerCanModify` was disabled so we couldn't push directly to this branch.

@daohoangson — if you'd prefer to keep this PR, you can cherry-pick the two fix commits from the `perf/structured-clone` branch on upstream:
```
git remote add upstream https://github.com/ag-ui-protocol/ag-ui.git
git fetch upstream perf/structured-clone
git cherry-pick cb80978b ad95de53
git push
```

Otherwise #1395 is ready to merge.

**Comment by jpr5** [comment]
Sorry for the delay, Gentlemen.  It's been merged.
